compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp

```diff
@@ -0,0 +1,240 @@
+  // Assumption: Layout is channel-last
+  if (!(channel_dim == add->dim(add->rank() - 1)))
+    return false;
+
+  mul = pred;
+  beta = constant;
+  return true;
+}
+
+} // namespace
+
+namespace luci
+{
+
+/**
+ *  Swap MUL/ADD if they are from batch normalization
+ *
+ *  BEFORE
+ *           [Mul]  gamma
+ *             |
+ *        [Add + Relu]  beta
+ *
+ *  AFTER
+ *           [Add]  beta/gamma
+ *             |
+ *           [Mul]  gamma
+ *             |
+ *           [Relu]
+ */
+bool swap_mul_add(luci::CircleAdd *add, std::vector<luci::CircleMul *> &mul_list,
+                  std::vector<luci::CircleAdd *> &add_list)
+{
+  luci::CircleNode *pred_node = nullptr;
+  luci::CircleMul *mul = nullptr;
+  luci::CircleConst *beta = nullptr;
+  luci::CircleConst *gamma = nullptr;
+
+  if (!is_batchnorm_add(add, mul, beta))
+    return false;
+
+  if (loco::succs(mul).size() != 1)
+    return false;
+
+  if (!is_batchnorm_mul(mul, pred_node, gamma))
+    return false;
+
+  if (beta->dtype() != loco::DataType::FLOAT32 || gamma->dtype() != loco::DataType::FLOAT32)
+    throw std::runtime_error("FusePreActivationBatchNormPass only supports Float32 model");
+
+  if (!is_non_negative(gamma))
+    return false;
+
+  // Insert Relu at the bottom
+  auto relu = add->graph()->nodes()->create<luci::CircleRelu>();
+  relu->features(mul);
+  loco::replace(add).with(relu);
+
+  // Replace beta <- beta / gamma
+  if (add->x() == beta)
+  {
+    add->y(pred_node);
+  }
+  else
+  {
+    add->x(pred_node);
+  }
+  add->fusedActivationFunction(luci::FusedActFunc::NONE);
+  uint32_t size = beta->size<loco::DataType::FLOAT32>();
+  for (uint32_t i = 0; i < size; ++i)
+  {
+    auto b = beta->at<loco::DataType::FLOAT32>(i);
+    auto g = gamma->at<loco::DataType::FLOAT32>(i);
+    if (b == g)
+    {
+      beta->at<loco::DataType::FLOAT32>(i) = 1;
+    }
+    else
+    {
+      // If g is 0, we use a small value (empirically determined)
+      if (g == 0)
+        g = 1e-10;
+      beta->at<loco::DataType::FLOAT32>(i) = b / g;
+    }
+  }
+
+  if (mul->x() == gamma)
+  {
+    mul->y(add);
+  }
+  else
+  {
+    mul->x(add);
+  }
+
+  mul_list.push_back(mul);
+  add_list.push_back(add);
+
+  return true;
+}
+
+bool FusePreActivationBatchNormPass::run(loco::Graph *g)
+{
+  LOGGER(l);
+  bool changed = false;
+
+  // Step 1. Swap MUL <-> ADD
+  for (auto node : loco::active_nodes(loco::output_nodes(g)))
+  {
+    auto add = dynamic_cast<luci::CircleAdd *>(node);
+    if (add == nullptr)
+      continue;
+
+    if (swap_mul_add(add, _mul_list, _add_list))
+      changed = true;
+  }
+
+  INFO(l) << "[FusePreActivationBatchNorm] Target pre-activations: " << _mul_list.size()
+          << std::endl;
+
+  // Valid pattern was not detected. Fast exit.
+  if (!changed)
+    return false;
+
+  // Step 2. Fuse MUL with the next CONV
+
+  // Step 3. Fuse ADD with the preceding CONV and insert SUB
+
+  // Step 4. Fuse SUB to CONV (SUB -> ADD <- CONV pattern)
+
```

Comment:
Just curiosity, I'm not sure but if the fusion proceeds sequentially, what if the condition that checks if it can keep going or not is fail? I mean, if Step3 returns false, what can we restore the graph?

Reply:
> I'm not sure but if the fusion proceeds sequentially, what if the condition that checks if it can keep going or not is fail? I mean, if Step3 returns false, how can we restore the graph?

The graph is sequentially changed as much as possible. It is not restored even if the next step fails. This is possible because each step is safe (= produces the same result after applied).